lib/mongo_client.js

```diff
@@ -504,6 +505,47 @@ MongoClient.prototype.startSession = function(options) {
   return this.topology.startSession(options, this.s.options);
 };
 
+/**
+ * Runs a given operation with an implicitly created session. The lifetime of the session
+ * will be handled without the need for user interaction.
+ *
+ * NOTE: presently the operation MUST return a Promise (either explicit or implicity as an async function)
+ *
+ * @param {Function} operation An operation to execute with an implicitly created session. The signature of this MUST be `(session) => {}`
+ * @param {Object} [options] Optional settings to be appled to implicitly created session
+ * @return {Promise} returns Promise if no callback passed
+ */
+MongoClient.prototype.withSession = function(operation, options, callback) {
+  if (typeof options === 'function') (callback = options), (options = undefined);
+  const session = this.startSession(options);
+
+  const cleanupHandler = (err, result, opts) => {
+    opts = Object.assign({ throw: true }, opts);
+    session.endSession();
+
+    if (typeof callback === 'function') {
+      return err ? callback(err, null) : callback(null, result);
+    } else {
+      if (err) {
+        if (opts.throw) throw err;
+        return Promise.reject(err);
+      }
+      return result;
+    }
+  };
+
+  try {
+    const result = operation(session);
+    const promise = isPromiseLike(result) ? result : Promise.resolve(result);
```

Comment:
if something is promiseLike, then `Promise.resolve` is a noop.

This can just be `const promise = Promise.resolve(result);`